dev.c

```diff
@@ -1135,7 +1158,175 @@ static void fuse_conn_queues_init(struct fuse_conn_queues *queue)
+	mm_segment_t old_fs = force_uaccess_begin();
+#else
+	mm_segment_t old_fs = get_fs();
+#endif
+
+	*mm_fault = false;
+	if (mm) {
+		if (!mmget_not_zero(mm)) {
+			*mm_fault = true;
+			return old_fs;
+		}
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,8,0)
+		use_mm(mm);
+		set_fs(USER_DS);
+#else
+		kthread_use_mm(mm);
+#endif
+	}
+
+	return old_fs;
+}
+
+void fuse_remove_user_access(struct mm_struct *mm, mm_segment_t old_fs)
+{
+	if (mm) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,8,0)
+		unuse_mm(mm);
+		set_fs(old_fs);
+#else
+		kthread_unuse_mm(mm);
+#endif
+		mmput(mm);
+	}
+}
+#else
+mm_segment_t fuse_setup_user_access(struct mm_struct *mm, bool *mm_fault) { *mm_fault = false; return get_fs();}
+void fuse_remove_user_access(struct mm_struct *mm, mm_segment_t old_fs) {}
+#endif
+
+static int fuse_process_user_queue(void *c)
+{
+	struct fuse_conn *fc = (struct fuse_conn*) c;
+	struct fuse_queue_cb *cb = &fc->queue->user_requests_cb;
+	bool mm_fault;
+	struct mm_struct *cur_mm = NULL;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,10,0)
+	mm_segment_t old_fs = force_uaccess_begin();
+#else
+	mm_segment_t old_fs = get_fs();
+#endif
+	unsigned long spin_wait = jiffies;
+
+	while (!kthread_should_stop()) {
+		do {
+			if (user_request_pending(fc)) break;
+			cpu_relax();
+		} while (jiffies < spin_wait);
+
+		// prepare to sleep
+		if (!user_request_pending(fc)) {
+			if (cur_mm != NULL) {
+				fuse_remove_user_access(cur_mm, old_fs);
+				cur_mm = NULL;
+			}
+
+			atomic_inc(&cb->r.need_wake_up);
+			if (signal_pending(current))
+				flush_signals(current);
+
+			wait_event_interruptible(fc->io_wait,
+				(user_request_pending(fc) || kthread_should_stop() ||
+				 kthread_should_park()));
+			atomic_dec(&cb->r.need_wake_up);
+		}
+
+		if (kthread_should_stop()) {
+			break;
+		}
+
+		if (kthread_should_park()) {
+			fuse_remove_user_access(cur_mm, old_fs);
+			cur_mm = NULL;
+			kthread_parkme();
+			continue;
+		}
+
+		mm_fault = false;
+		if (!cur_mm) {
+			fuse_setup_user_access(fc->user_mm, &mm_fault);
+			cur_mm = fc->user_mm;
+		}
+		fuse_run_user_queue(fc, mm_fault);
+		spin_wait = jiffies + msecs_to_jiffies(3);
+	}
+
+	if (cur_mm != NULL) {
+		fuse_remove_user_access(cur_mm, old_fs);
+	}
+	return 0;
+}
+
+void fuse_pause_user_queue(struct fuse_conn *fc)
+{
+	struct fuse_queue_cb *cb = &fc->queue->user_requests_cb;
+	int i;
+
+	pr_info("parking worker threads");
+	for (i=0; i<NWORKERS; i++)
+		if (fc->io_worker_thread[i]) kthread_park(fc->io_worker_thread[i]);
+	BUG_ON(atomic_read(&cb->r.need_wake_up) != 0);
+
+	if (fc->user_mm) {
+		mmdrop(fc->user_mm);
+		fc->user_mm = NULL;
+	}
+}
+
+void fuse_restart_user_queue(struct fuse_conn *fc)
+{
+	int i;
+
+	mmgrab(current->mm);
+	fc->user_mm = current->mm;
+	pr_info("unparking worker threads");
+	for (i=0; i<NWORKERS; i++)
+		if (fc->io_worker_thread[i]) kthread_unpark(fc->io_worker_thread[i]);
+}
+
+int fuse_conn_init(struct fuse_conn *fc, uint32_t max_workers)
```

Comment:
'max_workers's argument gets passed part of initialization. This is driven through a module param. Default is zero.